### PR TITLE
Add op_umask

### PR DESCRIPTION
**Pull request by dubiousjim**
This is part of a series of PRs towards #4017.

In other PRs we change permission tests for some other operations to check that the process's umask is being applied. There we assume the test machine is running with the usual default umask 0o022 (as Github's CI machines do); we could also query it using this function.

On Windows, `umask(value)` currently throws, but `umask()` just returns 0. Should we make the latter throw too?

### CODE REVIEW

**Comment by dubiousjim** [comment]
I need to (and will) add a unit test for this new functionality.

File: cli/fs.rs [hunk 1]
```diff
@@ -10,12 +10,40 @@ use rand;
...
+  Ok(previous.bits() as u32 & 0o777)
+}
+
+#[cfg(not(unix))]
+pub fn umask(mask: Option<u32>) -> Result<u32, ErrBox> {
+  match mask {
+    Some(_) => {
+      let e = std::io::Error::new(
+        std::io::ErrorKind::Other,
+        "Not implemented".to_string(),
```
Comment (ecyrbe) [inline cli/fs.rs]: maybe use [_umask](https://docs.microsoft.com/fr-fr/cpp/c-runtime-library/reference/umask?view=vs-2019) on windows like node does ?

File: cli/fs.rs [hunk 2]
```diff
@@ -10,12 +10,40 @@ use rand;
...
+      );
+      Err(ErrBox::from(e))
+    }
+    None => Ok(0),
+  }
+}
```
Comment (ry) [inline cli/fs.rs]: Can you move this code into the op_umask implementation itself? I don't think it's useful to have these separate methods in cli/fs.rs

File: cli/fs.rs (hunk 1 above)
Comment (ry) [reply cli/fs.rs]: Add a TODO with those links?

Comment (ry) [reply cli/fs.rs]: Use `OpError::other()`

File: cli/js/umask.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+/**
+ * **UNSTABLE**: maybe needs `allow-env` permissions.
+ *
+ * If `mask` is provided, sets the process umask. Always returns what the umask
+ * was before the call.
+ */
+export function umask(mask?: number): number {
+  return sendSync("op_umask", { mask });
+}
```
Comment (ry) [inline cli/js/umask.ts]: This belongs in `cli/js/ops/fs.ts` (which doesn't exist at time of writing; we're in the middle of a reorg). @bartlomieju where do you suggest he place this right now?

Also we need a test of this op...

File: cli/fs.rs (hunk 2 above)
Comment (dubiousjim) [reply cli/fs.rs]: I was going to do it later. As emerged in an earlier exchange, I wasn't clear on the intended division of labor between cli/fs.rs and cli/ops/fs.rs. So I implemented things in a way that I now realize you'd prefer to be folded into cli/ops/fs.rs. What I was planning to do was get this queue of PRs out for your review --- it's already complex keeping up with the changing master branch, waiting to push later PRs until earlier ones that block them have been merged, and so on. And then afterward, move all the stuff that doesn't need to be in cli/fs.rs into cli/ops/fs.rs. But if you prefer me to move stuff into cli/ops/fs.rs along the way, I'll do so.

File: cli/js/umask.ts (hunk 3 above)
Comment (dubiousjim) [reply cli/js/umask.ts]: test is coming...

Comment (bartlomieju) [reply cli/js/umask.ts]: > This belongs in `cli/js/ops/fs.ts` (which doesn't exist at time of writing; we're in the middle of a reorg). @bartlomieju where do you suggest he place this right now?
> 
> Also we need a test of this op...

`cli/js/ops/fs.rs` would be best, but I'm fine with landing as is - I'd rather reorg all `fs` ops in one go

**Comment by dubiousjim** [comment]
It's the `cafile_info` test that's failing now, and that's got nothing to do with this PR. Should be mergeable as is.

**Comment by dubiousjim** [comment]
Sorry to intermix my commits. It's still failing because now the `args` parameter is unused. Those unused variable warnings are really hard to squash when the function skips some functionality on Windows. (This came up in our exchange on another PR.) I think it needs either an ` #[allow(unused)]` somewhere, or some variable to take an underscore prefix, or two separate functions, one compiled on Windows the other on Unix.

**Comment by dubiousjim** [comment]
That's another option. :-)

**Comment by dubiousjim** [comment]
Great. I propose to now amend #4286 (tests and parts handling mkdir), #4287 (make_temp), 4289 (open/create) to use the explicit umask querying, and to return the same "not implemented" error you employed here if an explicit mode is requested on Windows. Let me know if you want to work on them yourself, or want to (make an executive decision for, not just express an inclination for) the noop behavior on Windows instead of throwing.

**Pull request merged**